source-mongodb/pull.go

```diff
@@ -204,6 +209,88 @@ type captureState struct {
 	StreamResumeToken bson.Raw `json:"stream_resume_token,omitempty"`
 }
 
+func migrateState(state *captureState, bindings []*pf.CaptureSpec_Binding) (bool, error) {
+	if state.Resources != nil && state.OldResources != nil {
+		return false, fmt.Errorf("application error: both Resources and OldResources were non-nil")
+	} else if state.Resources != nil {
+		log.Info("skipping state migration since it's already done")
+		return false, nil
+	}
+
+	state.Resources = make(map[boilerplate.StateKey]resourceState)
+
+	for _, b := range bindings {
+		if b.StateKey == "" {
+			return false, fmt.Errorf("state key was empty for binding %s", b.ResourcePath)
+		}
+
+		var res resource
+		if err := pf.UnmarshalStrict(b.ResourceConfigJson, &res); err != nil {
+			return false, fmt.Errorf("parsing resource config: %w", err)
+		}
+
+		ll := log.WithFields(log.Fields{
+			"stateKey":   b.StateKey,
+			"database":   res.Database,
+			"collection": res.Collection,
+			"resourceId": resourceId(res),
+		})
+
+		stateFromOld, ok := state.OldResources[resourceId(res)]
+		if !ok {
+			// This may happen if the connector has never emitted any checkpoints with data for this
+			// binding.
+			ll.Warn("no state found for binding while migrating state")
+			continue
+		}
+
+		state.Resources[boilerplate.StateKey(b.StateKey)] = stateFromOld
```

Comment:
nit: cast of StateKey isn't needed

Reply:
Hmm I'm finding that it is: `source-mongodb/pull.go:247:19: cannot use b.StateKey (variable of type string) as boilerplate.StateKey value in map index`

I may be missing something, but since the `migrateState` function is not intended to be permanent I'm not going to worry too much more about it right now.